tools/vendor/vendor-drift.py

```
@@ -0,0 +1,440 @@
+    if not components:
+        die_usage("the register declares no third-party components — nothing to "
+                  "diff, which is a broken register rather than a clean tree")
+    # Before the first network call, so a malformed register can never be
+    # mistaken for — or spend the clone budget of — an unreachable upstream.
+    for component in components:
+        absent = [k for k in REQUIRED_COMPONENT_KEYS if not component.get(k)]
+        if absent:
+            die_usage("component %s declares no %s — the register is malformed, "
+                      "which is not the same as an unreachable upstream"
+                      % (component.get("id") or component.get("path") or "<no id>",
+                         ", ".join(absent)))
```

Ruff (0.16.1)

[warning] Use format specifiers instead of percent format

Replace with format specifiers

(UP031)